lightning/src/onion_message/offers.rs

```diff
@@ -0,0 +1,181 @@
+	/// replying with an error.
+	fn handle_message(&self, message: OffersMessage) -> Option<OffersMessage>;
+}
+
+/// Possible BOLT 12 Offers messages sent and received via an [`OnionMessage`].
+///
+/// [`OnionMessage`]: crate::ln::msgs::OnionMessage
+#[derive(Debug)]
+pub enum OffersMessage {
+	/// A request for an [`Invoice`] for a particular [`Offer`].
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	InvoiceRequest(InvoiceRequest),
+
+	/// An [`Invoice`] sent in response to an [`InvoiceRequest`] or a [`Refund`].
+	///
+	/// [`Refund`]: crate::offers::refund::Refund
+	Invoice(Invoice),
+
+	/// An error from handling an [`OffersMessage`].
+	InvoiceError,
+}
+
+impl OffersMessage {
+	/// Returns whether `tlv_type` corresponds to a TLV record for Offers.
+	pub fn is_known_type(tlv_type: u64) -> bool {
+		match tlv_type {
+			INVOICE_REQUEST_TLV_TYPE | INVOICE_TLV_TYPE | INVOICE_ERROR_TLV_TYPE => true,
+			_ => false,
+		}
+	}
+
+	/// The TLV record type for the message as used in an `onionmsg_tlv` TLV stream.
+	pub fn tlv_type(&self) -> u64 {
+		match self {
+			OffersMessage::InvoiceRequest(_) => INVOICE_REQUEST_TLV_TYPE,
+			OffersMessage::Invoice(_) => INVOICE_TLV_TYPE,
+			OffersMessage::InvoiceError => INVOICE_ERROR_TLV_TYPE,
+		}
+	}
+
+	fn parse(tlv_type: u64, bytes: Vec<u8>) -> Result<Self, ParseError> {
+		match tlv_type {
+			INVOICE_REQUEST_TLV_TYPE => Ok(Self::InvoiceRequest(InvoiceRequest::try_from(bytes)?)),
+			INVOICE_TLV_TYPE => Ok(Self::Invoice(Invoice::try_from(bytes)?)),
+			INVOICE_ERROR_TLV_TYPE => Ok(Self::InvoiceError),
+			_ => unreachable!(),
+		}
+	}
+}
+
+impl Writeable for OffersMessage {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		match self {
+			OffersMessage::InvoiceRequest(message) => message.write(w),
+			OffersMessage::Invoice(message) => message.write(w),
+			OffersMessage::InvoiceError => Ok(()),
+		}
+	}
+}
+
+impl ReadableArgs<u64> for OffersMessage {
+	fn read<R: Read>(r: &mut R, tlv_type: u64) -> Result<Self, DecodeError> {
+		let mut bytes = Vec::new();
+		r.read_to_end(&mut bytes).unwrap();
+
+		match Self::parse(tlv_type, bytes) {
+			Ok(message) => Ok(message),
+			Err(ParseError::Decode(e)) => Err(e),
+			Err(_) => Err(DecodeError::InvalidValue), // TODO: log ParseError?
+		}
+	}
+}
+
+impl<M: Deref, T: Deref, ES: Deref, NS: Deref, SP: Deref, F: Deref, R: Deref, L: Deref>
+OffersMessageHandler for ChannelManager<M, T, ES, NS, SP, F, R, L>
+where
+	M::Target: chain::Watch<<SP::Target as SignerProvider>::Signer>,
+	T::Target: BroadcasterInterface,
+	ES::Target: EntropySource,
+	NS::Target: NodeSigner,
+	SP::Target: SignerProvider,
+	F::Target: FeeEstimator,
+	R::Target: Router,
+	L::Target: Logger,
+{
+	fn handle_message(&self, message: OffersMessage) -> Option<OffersMessage> {
+		match message {
+			OffersMessage::InvoiceRequest(invoice_request) => {
+				if !invoice_request.verify(&self.inbound_payment_key) {
```

Comment:
We simply won't be able to verify in this case. This is even more important when verifying an `Invoice` since we don't want to pay an `Invoice` that we haven't requested. Users would need to provide their own `OffersMessageHandler` if they want to do their own verification.